app/services/voucher_adjustments_service.rb

```diff
@@ -23,33 +23,35 @@ def self.calculate(order)
       handle_tax_included_in_price(order, amount)
     else
       adjustment.amount = amount
+      adjustment.save
     end
-
-    # Move to closed state
-    adjustment.close
   end
 
   def self.handle_tax_excluded_from_price(order, amount)
-    voucher_rate = amount / order.total
+    voucher_rate = amount / order.pre_discount_total
+
+    adjustment = order.voucher_adjustments.first
 
     # Adding the voucher tax part
     tax_amount = voucher_rate * order.additional_tax_total
 
-    adjustment = order.voucher_adjustments.first
     adjustment_attributes = {
-      amount: tax_amount,
       originator: adjustment.originator,
       order: order,
       label: "Tax #{adjustment.label}",
       mandatory: false,
-      state: 'closed',
+      state: 'open',
       tax_category: nil,
       included_tax: 0
     }
-    order.adjustments.create(adjustment_attributes)
+
+    # Update the amount if tax adjustment already exist, create if not
+    tax_adjustment = order.adjustments.find_or_initialize_by(adjustment_attributes)
+    tax_adjustment.amount = tax_amount
+    tax_adjustment.save
 
     # Update the adjustment amount
-    amount = voucher_rate * (order.total - order.additional_tax_total)
+    amount = voucher_rate * (order.pre_discount_total - order.additional_tax_total)
```

Comment:
You are re-assigning a variable given as a parameter here. I'm confused. I think the first is the voucher amount (already clamped to the order amount) and the second is the adjustment amount.

Just going through it in my head: If the voucher is more than the total there's additional tax of 10% then the voucher value will still be clamped to the pre-tax total. The voucher rate will be 100% which will create a tax adjustment to offset the tax completely. The order total will come down to zero. :heavy_check_mark:

Reply:
Indeed it's a bit confusing, I renamed it to `adjustment_amount`